### PR TITLE
Add basic check of dependencies

### LINKED ISSUE #21: Automatic Dependency Screening and Patching
## Problem Statement

Currently many common crates are not compatible with WASIX and require forks to work properly.

This results in a very suboptimal user experience.

`cargo wasix` should check for the presence of such dependencies in `Cargo.lock`, and either apply automatic fixes or inform the user.

This should apply to `cargo wasix {check,build,run}`.

## Implementation

`cargo wasix {check,build,run}` should:

* make sure the `Cargo.lock` has been generated
* Check the `Cargo.lock` for well-known incompatible crates
* In case a fork exists:
  automatically extend the `[patch.crates-io]` section with the appropriate patches
* In case no fork exists:
  Fail the build with a concrete error, like `crate XXX is not compatible with WASIX yet`

Care must be taken to pick branches that match the semver constraints for the patches, otherwise the patch will just be silently ignored.

## List of crates

* https://github.com/wasix-org/libc
* https://github.com/wasix-org/socket2
* https://github.com/wasix-org/tokio
* https://github.com/wasix-org/mio
* ...

## Example

A good example is the WASIX Axum example found here: http://wasix.org/docs/language-guide/rust/tutorials/wasix-axum

It requires the below Cargo.toml to work properly:

```toml
[package]
name = "wasix-axum"
version = "0.1.0"
edition = "2021"
 
[dependencies]
axum = { version = "=0.6.9", features = ["tokio"] } # ← Updated Line
tokio = { version = "=1.24.2", default-features = false, features = ["full"] } # ← Updated Line
parking_lot = { version = "=0.12.1", features = ["nightly"] } # ← Added Line
 
[patch.crates-io]
socket2 = { git = "https://github.com/wasix-org/socket2.git", branch = "v0.4.7" } # ← Added Line
libc = { git = "https://github.com/wasix-org/libc.git" } # ← Added Line
tokio = { git = "https://github.com/wasix-org/tokio.git", branch = "wasix-1.24.2" }
```

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
Very much a work in progress, but a start to fix #21.

Currently this prints something along the following lines when it finds an incompatible crates:
```
error: found incompatible crates in dependencies: ["mio"]
```

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
@theduke this is now ready for an initial review. It still only prints the dependencies that are known the be incompatible, but it now properly resolves dependencies of dependencies (e.g. it now logs libc and mio, instead of just mio). It now also has the known incompatible crates in `incompatible_crates/data.json`, which the tool automatically downloads and caches. Still to do:
* Refreshing of `incompatible_crates/data.json` after it's *X* duration old.
* Expanding the list of known incompatible crates.
* Either print suggested replacements, or automatically add it to `Cargo.toml` (might want to use a flag or the `fix` command for the latter).

File: src/lib.rs [hunk 1]
```diff
@@ -546,6 +558,172 @@ fn execute_cargo(cargo: &mut Command, config: &Config) -> Result<CargoBuild> {
...
+    repo: String,
+    /// Git branch to use.
+    branch: Option<String>,
+}
+
+fn known_incompatible_crates(config: &Config) -> Vec<IncompatibleCrate> {
+    match read_known_incompatible_crates(config) {
+        Ok(crates) => crates,
+        Err(err) => {
+            config.print_error(&err.context("not checking known incompatible crates"));
```
Comment (theduke) [inline src/lib.rs:591]: Good approach here to just print the error!

File: src/lib.rs [hunk 2]
```diff
@@ -546,6 +558,172 @@ fn execute_cargo(cargo: &mut Command, config: &Config) -> Result<CargoBuild> {
...
+    }
+}
+
+fn read_known_incompatible_crates(config: &Config) -> Result<Vec<IncompatibleCrate>> {
+    let mut path = Config::cache_dir()?;
+    path.push("incompatible_crates.json");
+
+    let file = match fs::File::open(&path) {
+        Ok(file) => file,
+        Err(ref err) if err.kind() == io::ErrorKind::NotFound => {
```
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/lib.rs]: Done. It now updates after 30 days.

**Comment by Thomasdezeeuw** [comment]
Ping @theduke

**Review by theduke — APPROVED** [review]

**Pull request merged**